src/tools/imgtool/library.cpp

```diff
@@ -128,155 +110,124 @@ void imgtool_library_add(imgtool_library *library, imgtool_get_info get_info)
-			next = m->next ? &m->next->previous : &library->last;
-			*previous = m->next;
-			*next = m->previous;
-			m->previous = nullptr;
-			m->next = nullptr;
-			return m;
-		}
-	}
-	return nullptr;
+	modulelist::iterator iter = find(module_name);
+	if (iter != m_modules.end())
+		m_modules.erase(iter);
 }
 
 
+//-------------------------------------------------
+//  module_compare
+//-------------------------------------------------
 
-static int module_compare(const imgtool_module *m1,
-	const imgtool_module *m2, imgtool_libsort_t sort)
+int library::module_compare(const imgtool_module *m1, const imgtool_module *m2, sort_type sort)
 {
 	int rc = 0;
 	switch(sort)
 	{
-		case ITLS_NAME:
-			rc = strcmp(m1->name, m2->name);
-			break;
-		case ITLS_DESCRIPTION:
-			rc = core_stricmp(m1->name, m2->name);
-			break;
+	case sort_type::NAME:
+		rc = strcmp(m1->name, m2->name);
+		break;
+	case sort_type::DESCRIPTION:
+		rc = core_stricmp(m1->description, m2->description);
+		break;
 	}
 	return rc;
 }
 
 
+//-------------------------------------------------
+//  sort
+//-------------------------------------------------
 
-void imgtool_library_sort(imgtool_library *library, imgtool_libsort_t sort)
+void library::sort(sort_type sort)
 {
-	imgtool_module *m1;
-	imgtool_module *m2;
-	imgtool_module *target;
-	imgtool_module **before;
-	imgtool_module **after;
-
-	for (m1 = library->first; m1; m1 = m1->next)
+	auto compare = [&](const std::unique_ptr<imgtool_module> &a, const std::unique_ptr<imgtool_module> &b)
 	{
-		target = m1;
-		for (m2 = m1->next; m2; m2 = m2->next)
-		{
-			while(module_compare(target, m2, sort) > 0)
-				target = m2;
-		}
-
-		if (target != m1)
-		{
-			/* unlink the target */
-			before = target->previous ? &target->previous->next : &library->first;
-			after = target->next ? &target->next->previous : &library->last;
-			*before = target->next;
-			*after = target->previous;
-
-			/* now place the target before m1 */
-			target->previous = m1->previous;
-			target->next = m1;
-			before = m1->previous ? &m1->previous->next : &library->first;
-			*before = target;
-			m1->previous = target;
-
-			/* since we changed the order, we have to replace ourselves */
-			m1 = target;
-		}
-	}
+		return module_compare(a.get(), b.get(), sort) < 0;
+	};
+	m_modules.sort(compare);
 }
 
 
+//-------------------------------------------------
+//  find
+//-------------------------------------------------
 
-const imgtool_module *imgtool_library_findmodule(
-	imgtool_library *library, const char *module_name)
+library::modulelist::iterator library::find(const char *module_name)
 {
-	const imgtool_module *module;
-
-	assert(library);
-	module = library->first;
-	while(module && module_name && strcmp(module->name, module_name))
-		module = module->next;
-	return module;
+	return std::find_if(
+		m_modules.begin(),
+		m_modules.end(),
+		[&](std::unique_ptr<imgtool_module> &module) { return !strcmp(module->name, module_name); });
```

Comment:
Please don't use auto-capture - explicitly capture `module_name` to make things clear.